modepy/matrices.py

```diff
@@ -252,33 +252,68 @@ def __init__(self, face_vertices):
         if npts != vol_dim:
             raise ValueError("face_vertices has wrong shape")
 
-        self.origin = face_vertices[:, 0]
-        self.span = face_vertices[:, 1:] - self.origin[:, np.newaxis]
+        self.origin = face_vertices[:, 0].reshape(-1, 1)
+        self.span = face_vertices[:, 1:] - self.origin
 
         self.face_dim = vol_dim - 1
 
     def __call__(self, points):
-        return (self.origin[:, np.newaxis]
-                + np.einsum("ad,dn->an", self.span, points*0.5 + 0.5))
+        return self.origin + np.einsum("ad,dn->an", self.span, points*0.5 + 0.5)
 
 
-def modal_face_mass_matrix(trial_basis, order, face_vertices, test_basis=None):
+class _HypercubeFaceMap:
+    def __init__(self, face_vertices):
+        """
+        :arg face_vertices: an array of shape ``[dim, npts]``, where *npts*
+            should equal `2**(dim - 1)`.
+        """
+        vol_dim, npts = face_vertices.shape
+        if npts != 2**(vol_dim-1):
+            raise ValueError("face_vertices has wrong shape")
+
+        self.origin = face_vertices[:, 0].reshape(-1, 1)
+        self.span = face_vertices[:, -2:0:-1] - self.origin
+
+        self.face_dim = vol_dim - 1
+
+    def __call__(self, points):
+        return self.origin + np.einsum("ad,dn->an", self.span, points*0.5 + 0.5)
+
+
+def modal_face_mass_matrix(trial_basis, order, face_vertices,
+        test_basis=None, domain=None):
     """
     :arg face_vertices: an array of shape ``[dim, npts]``, where *npts*
         should equal *dim*.
+    :arg domain: identifier for the reference element, can be one of
+        `"simplex"` or `"hypercube"`.
```

Comment:
I used this "domain" thing before, but I'm not exactly sure it's a good API. Any thoughts/suggestions?

Reply:
How about this: Make `Hypercube` and `Simplex` classes somewhere that know their dimension (but nothing else) (`modepy.shapes`?), and then make `get_face_map` and `get_quadrature_with_order` and use `functools.singledispatch` to register implementations? Over time, we could use this pattern in more of modepy. Julia: the good parts. :)

Reply:
Stubbed something out along those lines in 704e686. What do you think?

I put a `get_basis` (and friends) in there that feel like they belong in `modepy.modes` (or `modepy.nodes`, etc) where the things are actually defined. Should I move them?

Reply:
Yeah, I'd feel better about them being moved. One advantage of the `singledispatch` solution is that these functions don't need to live in the same file as the class.

Reply:
Moved most things out in 7acf2b0.